### ISSUE TITLE
Better hot reloading workflow?

### ISSUE THREAD

**Issue by geekrelief** [geekrelief/gdnim#27]
While working with hot reloading here are some issues I've come across and some ideas on how to improve the workflow.

- Serializing - Can we make reloading data easier by serializing all the data in the component?
  - Currently, you need to `save()` before you can `load()`, and often you don't know what you need to save before you reload.
  - Can we uses `inst2dict` for this?

- Handling crashes / nil access - Can prevent a bad component from taking down the whole app? 
  - When recompiling a component, an error like nil access can occur which takes down the app.
  - Ideally, the error would be caught, the component unloaded, and the app paused.
  - Watcher would then just wait until you recompile and attempt to reload and continue.
  
- Dependencies - Can we make it easier to clear references to reloading nodes?
  - Maybe a macro can be created to`connect` to a node see if it's reloading.  That would require the reloading node to emit a signal to distinguish it from exiting the tree normally. 
  - Right now the watcher emits a signal when an instance reloads, but anything that binds to it will be receiving the signal for all instances and have to filter for the node it cares about.
  - Unloading is easy, but what about reloading the references? Watcher emits a signal for the dependents to reload their dependency references.

**Comment by geekrelief** [comment]
With 0c76a6f, Watcher automatically serializes any properties of a component from `getPropertyListImpl()`. Also anything with `{.gdExport.}` is saved as long as it is not of VariantType.Object.  The `save()/load()` macros are still useful for anything that is an object or stores references to stuff. You need to handle the serialization yourself.

**Comment by geekrelief** [comment]
It should be easier to recursively serialize a hierarchy of components now that we save properties.

**Comment by geekrelief** [comment]
On dealing with nil accesses, can we parse the code to see if a reference has been initialized before it is used? For each proc, we place code just before the first use of a reference to see if it's nil.

**Comment by geekrelief** [comment]
Added a macro `ifValid` which can be used like:
```nim
  ifValid self.tween:
    self.tween.start()
```
This will prevent the app from crashing by doing a nil check. And prints out a message to console if nil is detected. The message can be toggled with the `verbose_nil_check` setting in the `build.ini`.

**Comment by geekrelief** [comment]
I'm going to be changing the semantics of the `reload` section and `dependencies`.  So don't depend on not having to define a `ready` method.  I figured a way to pack an entire hierarchy with PackedScene, but you probably still want to use `save()/load()` macro for certain things that aren't capturable as a property. In some situations you want some initialization to occur in `enter_tree` and some in `ready`, depending on if you're adding children, connecting signals, or getting references to other nodes.  So I'm thinking the `reload` section should only be for data initialization which will run at the top of `enter_tree`. With hierarchy packing, nodes no longer need to free their references since Watcher will destroy them.  A component just needs to declare its dependencies, and initialization code can be placed in `enter_tree` or `ready`.

**Comment by geekrelief** [comment]
Got scene hierarchy reloading with custom data saving rough draft working. Just need to do more testing and cleanup.

**Comment by geekrelief** [comment]
After more testing, I had to rewrite how the scene state reloading works.  I make use of a single PackedScene to save the state of any affected component instances by removing them from the scene tree and putting them into a new node for packing, then updating the library and instancing the saved nodes.  This new approach works in parallel with the custom instance data.

The `dependencies` section semantics have been simplified, so now you just need to declare which components a component depends on.  The initialization code can be moved to `enter_tree` or `ready` when interacting with dependencies.  In `enter_tree` or `ready`, you can check if the comp is reloading with `isNotReloading()` and initialize references, `connect`, `add_child`, etc.

I still need to test with the `reload` switch turned off. Then clean up the API and code for maintenance.

**Comment by geekrelief** [comment]
`reload` off works fine.  Hot reloading is more robust now.  We can save all the state with a combination of PackedScenes and serialization. There are also procs you can call to check if the app is reloading, so you add children at the start, but avoid adding on reload.  We can also change the type of a component (.nim and .tscn) and have it reload properly, for example if you have a component of type `HBoxContainer` you can hot reload with a `VBoxContainer` if the nim type and tscn are updated.

**Comment by geekrelief** [comment]
Tested workflow.  Merged into master.
